tests/tests_pytorch/callbacks/test_weight_averaging.py

```diff
@@ -0,0 +1,288 @@
+            assert self._average_model.module._trainer is None
+            self.setup_called = True
+
+    callback = TestCallback()
+    trainer = Trainer(default_root_dir=tmp_path, callbacks=callback, fast_dev_run=True)
+    trainer.fit(BoringModel(), train_dataloaders=DataLoader(RandomDataset(32, 2)))
+    assert callback.setup_called
+
+
+@pytest.mark.parametrize("batch_norm", [True, False])
+@pytest.mark.parametrize("iterable_dataset", [True, False])
+def test_ema(tmp_path, batch_norm: bool, iterable_dataset: bool):
+    _train(tmp_path, EMATestCallback(), batch_norm=batch_norm, iterable_dataset=iterable_dataset)
+
+
+@pytest.mark.parametrize(
+    "accelerator", [pytest.param("gpu", marks=RunIf(min_cuda_gpus=1)), pytest.param("mps", marks=RunIf(mps=True))]
+)
+def test_ema_accelerator(tmp_path, accelerator):
+    _train(tmp_path, EMATestCallback(), accelerator=accelerator, devices=1)
+
+
+@RunIf(min_cuda_gpus=2, standalone=True)
+def test_ema_ddp(tmp_path):
+    _train(tmp_path, EMATestCallback(devices=2), strategy="ddp", accelerator="gpu", devices=2)
+
+
+@RunIf(min_cuda_gpus=2)
+def test_ema_ddp_spawn(tmp_path):
+    _train(tmp_path, EMATestCallback(devices=2), strategy="ddp_spawn", accelerator="gpu", devices=2)
+
+
+@RunIf(skip_windows=True)
+def test_ema_ddp_spawn_cpu(tmp_path):
+    _train(tmp_path, EMATestCallback(devices=2), strategy="ddp_spawn", accelerator="cpu", devices=2)
+
+
+@pytest.mark.parametrize("crash_on_epoch", [1, 3])
+def test_ema_resume(tmp_path, crash_on_epoch):
+    _train_and_resume(tmp_path, crash_on_epoch=crash_on_epoch)
+
+
+@RunIf(skip_windows=True)
+def test_ema_resume_ddp(tmp_path):
+    _train_and_resume(tmp_path, crash_on_epoch=3, use_ddp=True)
+
+
+def test_swa(tmp_path):
+    _train(tmp_path, SWATestCallback())
+
+
+def _train(
+    tmp_path: str,
+    callback: WeightAveraging,
+    batch_norm: bool = True,
+    strategy: str = "auto",
+    accelerator: str = "cpu",
+    devices: int = 1,
+    iterable_dataset: bool = False,
+    checkpoint_path: Optional[str] = None,
+    crash_on_epoch: Optional[int] = None,
+) -> None:
+    trainer = Trainer(
+        default_root_dir=tmp_path,
+        enable_progress_bar=False,
+        enable_model_summary=False,
+        logger=False,
+        max_epochs=8,
+        num_sanity_val_steps=0,
+        callbacks=callback,
+        accumulate_grad_batches=2,
+        strategy=strategy,
+        accelerator=accelerator,
+        devices=devices,
+    )
+    model = WeightAveragingTestModel(
+        batch_norm=batch_norm, iterable_dataset=iterable_dataset, crash_on_epoch=crash_on_epoch
+    )
+
+    if crash_on_epoch is None:
+        trainer.fit(model, ckpt_path=checkpoint_path)
+    else:
+        with pytest.raises(Exception, match="CRASH TEST"):
+            trainer.fit(model, ckpt_path=checkpoint_path)
+
+    assert trainer.lightning_module == model
+
+
+def _train_and_resume(tmp_path: str, crash_on_epoch: int, use_ddp: bool = False) -> None:
```

Comment:
This tests that we can crash and resume, but afaict it doesn't test whether the resulting averaging is equivalent. We can harden this in a subsequent PR, but it is important to know for sure that averaging works if I stop training and resume while averages are being taken, irrespective of where I stop and resume in the lifecycle.

Reply:
I'll try to still improve the test.

Reply:
@lantiga now I test that after stopping and resuming we get the same final model. The parameters are not identical - I have to use `atol=0.001` - but they are close enough so that I think that the difference comes from some random change instead of a bug in restoring the checkpoint. I don't know what could cause the difference, though. I pass `deterministic=True` to Trainer. I'm curious if you have some ideas, or if you think that that's close enough.